### ISSUE TITLE
3.0 read possible read regression - CPU essentially idle during timeouts

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#3835]
I have just reverted my system to 2.3 and confirmed that this is a regression in 3.0

The workload is 2 clients, one of them does:

```
$STRESS read duration=3h -mode cql3 native user=cassandra password=cassandra \
	 -rate threads=600 limit=400000/s  -pop 'dist=uniform(1..30000000)' \
	 -col n='FIXED(1)' size='FIXED(4096)' -node 40.40.40.1 -errors ignore
```

(after ingesting that many keys with those sizes above)

In parallel, I run a full table scan from our code examples:

```
go run ./efficient_full_table_scan.go --username nolat --password 1234 -s 40 -c 86 -n 1 40.40.40.1
```

I am disabling the cache to make this easier to reproduce.
Also disabling `mc` to narrow this down.

The behavior in 2.3 is that I see timeouts in the client side, but grafana never registers semaphore exceptions. The load is still going on, despite slower with the occasional timeout. The CPUs are at around 100% useful load.

In 3.0, both the range scan and the cassandra-stress work timeout all the time, and the cassandra-stress grinds to a halt. The logs are filled with semaphore timed out messages, but what struck me as odd, is that all the CPUs (as seen by scylla) are in 0%. So they are essentially not doing anything.

Any idea why? Anything I can look for? physical access can be granted if needed.

2.3 (in the left) vs 3.0 (in the right) interesting grafana graphs per shard
![f](https://user-images.githubusercontent.com/331197/46690514-7fa5ab80-cbd0-11e8-9074-836e969b9cae.png)

![r](https://user-images.githubusercontent.com/331197/46690528-83d1c900-cbd0-11e8-9ec3-2022cb4bd7cf.png)

(in this one, notice how the reads essentially stop happening altogether)
![p](https://user-images.githubusercontent.com/331197/46690457-5be26580-cbd0-11e8-9f55-9af45261f5a0.png)

**Comment by glommer** [comment]
@denesb that looks like your turf

**Comment by denesb** [comment]
How many nodes? How many shards?

> The logs are filled with semaphore timed out messages

Do you mean the semaphore from the restricted reader?

> but what struck me as odd, is that all the CPUs (as seen by scylla) are in 0%. So they are essentially not doing anything.

I wonder if cached readers are sitting on permits. We have code that prevents that, but nevertheless...
Can you share the graphs for the following metrics:
* `scylla_database_querier_cache_population` (for this we are interested in the absolute value, don't irate it).
* `scylla_database_querier_cache_resource_based_evictions`

Another possible answer is that the multishard reader goes nuts with read-ahead. Unfortunately there is no metrics to verify that directly. Let's confirm first that the permit (resource) based eviction works correctly.

**Comment by glommer** [comment]
@denesb 

> Do you mean the semaphore from the restricted reader?

I wish I knew for sure, the log messages don't say. They just say semaphore timed out. Traditionally that has been the restricted reader.

>I wonder if cached readers are sitting on permits.

Sounds like a plausible theory.

> scylla_database_querier_cache_population

![pop by shard](https://user-images.githubusercontent.com/331197/46758846-dd52fa00-cc9b-11e8-9c5a-cc2eb01b67e7.png)

> scylla_database_querier_cache_resource_based_evictions

![rbe](https://user-images.githubusercontent.com/331197/46758952-2440ef80-cc9c-11e8-92c9-3323b654106a.png)

**Comment by denesb** [comment]
After some playing around these are my first conclusions and observations:
`efficient_full_table_scan.go` runs with a parallelism of 3440, and it creates a cql session for each thread. Some of the sessions fail to open (not even the TCP connection is set up due to timeout), but still this results in around ~10k connections to the server (1 node, 86 shards). According to my observation this is the source of the symptoms we are seeing.
I managed to reproduce the meltdown with `scylla_bench` (patched to scan the right table). When the connection count is increased to ~2500 it starts to produce symptoms similar to those mentioned above.

I have no problem running a range scan parallel to single partition reads, when the range scans are run with `scylla-bench`:
```
./scylla-bench -mode=scan -concurrency=258 -range-count=25800 -connection-count=258 -nodes=40.40.40.1 --username=nolat --password=1234 -keyspace=keyspace1 -table=standard1
```
The load stabilizes around ~50%, handling 200k OPS, not a single timeout.

**Comment by denesb** [comment]
Increasing concurrency tenfold causes Scylla to start trashing:
```
./scylla-bench -mode=scan -concurrency=2580 -range-count=258000 -connection-count=258 -nodes=40.40.40.1 --username=nolat --password=1234 -keyspace=keyspace1 -table=standard1
```
So the cause is compound: too high parallelism and too high connection count. @glommer is a node this size supposed to handle such load? Regardless of the answer I think we should handle overload more gracefully than completely dropping the ball. Maybe after a certain point we should reject new messages at the transport layer.

**Comment by denesb** [comment]
Ok I just read the initial description again, 2.3 can handle this load gracefully. So there is a regression her.

**Comment by denesb** [comment]
I think I know what the problem is.

Previously, the range scan was sending individual requests to shards. These requests created a reader from scratch and destroyed it after they were done.

The new range scan code creates a `multishard_combining_reader` which keeps *all* its shard readers alive for the duration of the read.

That is, the previous algorithm only had as much readers alive as much was needed in the current iteration, while the new one keeps all reader alive, even if they were used only once (and are not needed again in the page). There are tradeoffs here: the old algorithm had increased read amplification, due to the repeated creation-destoyal cycle (and all that comes with it). The new algorithm has a lot more active readers. This is made worse by read-ahead.

The practical implications of this is that when the concurrency of the scans is high enough (like in the case reported in this issue), our sstable reader queues fill up and the semaphores start to time out. A single shard reader being killed by timeout will kill the entire scan. So none of the reads can make forward progress, Scylla starts trashing.

To quickly illustrate my point, this is the queued sstable stats for with exponentially increasing read-ahead, linearly increasing read-ahead and no read-ahead respectively:
![image](https://user-images.githubusercontent.com/1389273/46866338-add1f800-ce29-11e8-8c95-b77732d577b3.png)
(I stopped the bit with the linearly increasing read-ahead too early, messing up the graphs a bit)

**Comment by denesb** [comment]
I'm not sure what to propose as solution at this point.

Making read-ahead less aggressive, or even disabling it helps, but it doesn't go all the way. BTW I added some counters and it seems that all readers created as part of a read-ahead end up actually contributing to the page content (at least for this workload). Normally even if the read ahead isn't used, it is highly likely that it will be used on the next page, given that the saved state survives.
A side effect of this trashing is that the querier cache is empty almost all the time, as inserted readers are immediately evicted to free up a permit.

**Comment by glommer** [comment]
@tzach @slivne this is a confirmed 3.0 regression.

@denesb I think we should just not account those readers as active sstable reads. The purpose of limiting active sstable reads is to limit memory usage, in which case we should probably just limit memory usage in some other way in parallel to that, that is exclusive to range scans.

Even without the time outs, if we use up all of the active sstable permits workloads that are running concurrently to the range scan will suffer increased latencies as we are effectively reducing their opportunity to read.

Repairs had a similar problem and we had to use a separate set of active sstables for them

**Comment by glommer** [comment]
> 
> The practical implications of this is that when the concurrency of the scans is high enough (like in the case reported in this issue), our sstable reader queues fill up and the semaphores start to time out. A single shard reader being killed by timeout will kill the entire scan. So none of the scans can make forward progress, Scylla starts trashing.

It's much worse than "the scans can't make progress". Other read workloads that are not scanning make no progress either.

Doing online work and a daily /weekly range scan is a very common scenario, to be made even more common by our per-user SLA feature.

**Comment by denesb** [comment]
Having a separate concurrency controlling semaphore sounds like a good start. I think that we also need admission control. Contrary to singlular partition queries, where there is a single permit to obtain (so a query either succeeds or fails), ranges scans create shard readers on-demand, so they can read 90% of the required data and get a timeout at the last shard, failing the entire query, wasting all that work. Semaphore timeouts are very dangerous to range scans, we should only admit them if they have a reasonable chance of finishing.

**Comment by avikivity** [comment]
@denesb new reads shouldn't have to wait for the cache. But we have this in wait_admission():

```
    if (!may_proceed(r) && _evict_an_inactive_reader) {
        while (_evict_an_inactive_reader() && !may_proceed(r));
    }
```

Why doesn't it evict inactive readers?

**Comment by slivne** [comment]
Not related - yet can we detect this case on the field.

We have separate timeout counters for range scans and regular reads, but do we have separate counters for requests so we can detect now many range scans are not making progress vs regular reads ?

**Comment by denesb** [comment]
@avikivity  there are no inactive readers to evict. The cache is effectively empty, inserted readers are almost instantaneously evicted.

**Comment by denesb** [comment]
@slivne we have counters for reads and timed out reads. But we don't count range scans separately.

**Comment by denesb** [comment]
So I added a separate semaphore for range-scans and it appears to solve the problem. Normal reads continue to work undisturbed (without a single timeout) even when range scans go trashing. This of course comes at the cost of the memory consumption of the additional queue. For now the range-scan queue has the same parameters as that of the normal reads (2% of the shard's memory, 100 active readers max). This effectively doubles the resources we reserve for active/inactive reads. Should we reserve less for range scans? No matter how little we reserve, it will still be on top of what we had until now, unless we decrease the existing reserves for normal reads.

The real amount of resource we allocate for these queues is becoming less and less clear with every new queue we add. A real solution would be a shared queue which knows about priorities (#3851).

**Comment by denesb** [comment]
To solve the trashing we need admission control.

**Comment by denesb** [comment]
I've opened a separate issue for admission control: #3856.
Let's solve the regression of affecting normal reads first, as its more severe.

**Comment by tgrabiec** [comment]
This issue is another example of why `efficient_full_table_scan.go` is not as efficient as it could be: https://github.com/scylladb/scylla/issues/3710#issuecomment-417001289
